### PR TITLE
feat(cf): Fetch created service key via SpEL

### PR DESCRIPTION
**Pull request by stuart-pollock**
spinnaker/spinnaker#4260

### CODE REVIEW

File: orca-core/src/test/java/com/netflix/spinnaker/orca/pipeline/expressions/functions/ServiceKeyExpressionFunctionProviderTest.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+  @Test
+  void serviceKeyShouldReturnEmptyMapForNoResult() {
+    Map<String, Object> katoTaskMapWithoutResults = new ImmutableMap.Builder<String, Object>()
+      .put("id", "task-id")
+      .put("status", SUCCEEDED)
+      .put("history", Collections.emptyList())
+      .put("resultObjects", Collections.emptyList())
+      .build();
+    Execution execution = new Execution(PIPELINE, "stage-name-1", "application-name");
+    Map<String, Object> contextWithoutServiceKey = new HashMap<>();
```
Comment (marchello2000) [inline orca-core/src/test/java/com/netflix/spinnaker/orca/pipeline/expressions/functions/ServiceKeyExpressionFunctionProviderTest.java]:
```
    Map<String, Object> contextWithoutServiceKey = new HashMap<>();
    contextWithoutServiceKey.put("parentStage", "parent-stage");
    contextWithoutServiceKey.put("account", "account-name");
    contextWithoutServiceKey.put("credentials", "my-account");
    contextWithoutServiceKey.put("region", "org > space");
```
is repeated a bunch of times, can make a helper function to make test more readable

File: orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/expressions/functions/ServiceKeyExpressionFunctionProvider.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        new FunctionParameter(String[].class, "ids", "A list of stage name or stage IDs to search")
+      ))
+    );
+  }
+
+  public static Map<String, Object> serviceKey(Execution execution, String... id) {
+    return execution.getStages()
+      .stream()
+      .filter(matchesServiceKeyStage(id))
+      .findFirst()
```
Comment (marchello2000) [inline orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/expressions/functions/ServiceKeyExpressionFunctionProvider.java]: can there only be one `createServiceKeyStage`?

File: orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/expressions/functions/ServiceKeyExpressionFunctionProvider.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+            }
+          });
+
+          return serviceKeyDetails;
+        })
+      .orElse(Collections.emptyMap());
+  }
+
+  private static Predicate<Stage> matchesServiceKeyStage(String... id) {
+    List<String> idsOrNames = Arrays.asList(id);
```
Comment (marchello2000) [inline orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/expressions/functions/ServiceKeyExpressionFunctionProvider.java]: can/should be `final`

File: orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/expressions/functions/ServiceKeyExpressionFunctionProvider.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+  @Override
+  public String getNamespace() {
+    return null;
+  }
+
+  @NotNull
+  @Override
+  public Collection<FunctionDefinition> getFunctions() {
+    return Collections.singletonList(
+      new FunctionDefinition("serviceKey", Arrays.asList(
```
Comment (marchello2000) [inline orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/expressions/functions/ServiceKeyExpressionFunctionProvider.java]: my concern here is that this is the first function that is (sort of) provider specific. Perhaps worth prefixing it with `cf`, e.g. `cfServiceKey`? cc: @robzienert

Comment (marchello2000) [reply orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/expressions/functions/ServiceKeyExpressionFunctionProvider.java]: Also, don't forget to document it here: https://github.com/spinnaker/spinnaker.github.io/blob/master/reference/pipeline/expressions/index.md

Comment (robzienert) [reply orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/expressions/functions/ServiceKeyExpressionFunctionProvider.java]: This was actually the point of to `getNamespace` method. So cloud provider spel helpers could be prefixed with the provider name. Then calling the method would look like so:

```
${#cloudfoundry.serviceKey(...)}
```

All this said, I don't like our provider specific logic in `orca-core` but that's currently the what we do for the other spel helpers today so I figure something out for think it's fine for now. This seems a perfect candidate later on for plugins?

File: orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/expressions/functions/ServiceKeyExpressionFunctionProvider.java (hunk 3 above)
Comment (stuart-pollock) [reply orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/expressions/functions/ServiceKeyExpressionFunctionProvider.java]: IntelliJ complains:  `As private methods cannot be meaningfully overridden, declaring them final is redundant.`.  We've been suggested to follow IntelliJ's suggestions in the past.  It's also consistent with https://github.com/spinnaker/orca/blob/820148ef29850cdce320aa0af3cad484065f4989/orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/expressions/functions/DeployedServerGroupsExpressionFunctionProvider.java#L93.

File: orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/expressions/functions/ServiceKeyExpressionFunctionProvider.java (hunk 2 above)
Comment (stuart-pollock) [reply orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/expressions/functions/ServiceKeyExpressionFunctionProvider.java]: There may be multiple `createServiceKey` stages.  Each will have its own name, and we expect them to be addressed primarily by name rather than by id.  We have removed the spread operator.

File: orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/expressions/functions/ServiceKeyExpressionFunctionProvider.java (hunk 3 above)
Comment (marchello2000) [reply orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/expressions/functions/ServiceKeyExpressionFunctionProvider.java]: i meant the `idsOrNames`

Comment (stuart-pollock) [reply orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/expressions/functions/ServiceKeyExpressionFunctionProvider.java]: We've eliminated the variable in response to another of your suggestions.

File: orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/expressions/functions/ServiceKeyExpressionFunctionProvider.java (hunk 4 above)
Comment (ajordens) [reply orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/expressions/functions/ServiceKeyExpressionFunctionProvider.java]: +1 on the namespacing, that was the intention.

File: orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/expressions/functions/ServiceKeyExpressionFunctionProvider.java (hunk 2 above)
Comment (marchello2000) [reply orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/expressions/functions/ServiceKeyExpressionFunctionProvider.java]: got it, so the above should be `"id"` instead of `"ids"` (in the params)

**Comment by marchello2000** [comment]
looks good, just add the namespace and change `ids` to `id`

File: orca-core/src/test/java/com/netflix/spinnaker/orca/pipeline/expressions/ExpressionTransformTest.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    public String getNamespace() {
+      return "testNamespace";
+    }
+
+    @NotNull
+    @Override
+    public Collection<FunctionDefinition> getFunctions() {
+      return Collections.singletonList(
+        new FunctionDefinition("testFunctionInNamespace", singletonList(
+          new FunctionParameter(Execution.class, "functionParam", "Function param")
```
Comment (marchello2000) [inline orca-core/src/test/java/com/netflix/spinnaker/orca/pipeline/expressions/ExpressionTransformTest.java]: from the usage of how you have it above, this should be:
```
Arrays.asList(
        new FunctionParameter(Execution.class, "execution", "Execution param"),
        new FunctionParameter(String.class, "functionParam", "Function param")
))
```

instead of `singletonList`

File: orca-core/src/test/java/com/netflix/spinnaker/orca/pipeline/expressions/ExpressionTransformTest.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+  void includeExecutionParameterDoesNotInsertNullNamespaceInExpressionFunctionProvider() {
+    String expression = "${#testFunctionNoNamespace(\"param\")}";
+    String expectedResult = "${#testFunctionNoNamespace( #root.execution, \"param\")}";
+
+    String result = expressionTransform.includeExecutionParameter(expression);
+
+    assertThat(result).isEqualTo(expectedResult);
+  }
+
+  private static class TestExpressionFunctionProviderWithNamespace implements ExpressionFunctionProvider {
```
Comment (marchello2000) [inline orca-core/src/test/java/com/netflix/spinnaker/orca/pipeline/expressions/ExpressionTransformTest.java]: nit: you could fold `TestExpressionFunctionProviderWithNamespace ` and `TestExpressionFunctionProviderWithoutNamespace` into a single class and then pametrize the `namespace` and the `params`

File: orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/expressions/ExpressionTransform.java [hunk 7]
```diff
@@ -270,21 +270,31 @@ static String escapeSimpleExpression(String expression) {
...
-  private String includeExecutionParameter(String e) {
+  // Visible for testing
+  String includeExecutionParameter(String e) {
     String expression = e;
 
     // An expression aware function is any that takes an Execution as its first parameter
     Set<String> expressionAwareFunctions = new HashSet<>(EXECUTION_AWARE_FUNCTIONS);
-    expressionFunctionProviders.forEach(p -> {
-      p.getFunctions().forEach(f -> {
+    for (ExpressionFunctionProvider p : expressionFunctionProviders) {
```
Comment (marchello2000) [inline orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/expressions/ExpressionTransform.java]: i guess `namespace` was only implemented half-way, thank you for fixing it

**Comment by marchello2000** [comment]
LGTM, but see if @ajordens has any other comments.

File: orca-core/src/test/java/com/netflix/spinnaker/orca/pipeline/expressions/functions/ServiceKeyExpressionFunctionProviderTest.java [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+import static com.netflix.spinnaker.orca.ExecutionStatus.RUNNING;
+import static com.netflix.spinnaker.orca.ExecutionStatus.SUCCEEDED;
+import static com.netflix.spinnaker.orca.pipeline.model.Execution.ExecutionType.PIPELINE;
+import static java.util.Collections.emptyList;
+import static java.util.Collections.singletonList;
+import static java.util.Collections.singletonMap;
+import static org.assertj.core.api.Assertions.assertThat;
+
+class ServiceKeyExpressionFunctionProviderTest {
```
Comment (ajordens) [inline orca-core/src/test/java/com/netflix/spinnaker/orca/pipeline/expressions/functions/ServiceKeyExpressionFunctionProviderTest.java]: For future reference, I still believe we can author tests in groovy (we just would prefer non-tests be java/kotlin).

Unit tests in strict java are still a little verbose for my liking (don't need to do this as part of this PR but consider it next time).

**Comment by ajordens** [comment]
LGTM after the namespace `_` vs `.` clarification.

File: orca-core/src/test/java/com/netflix/spinnaker/orca/pipeline/expressions/ExpressionTransformTest.java (hunk 5 above)
Comment (stuart-pollock) [reply orca-core/src/test/java/com/netflix/spinnaker/orca/pipeline/expressions/ExpressionTransformTest.java]: 👍

File: orca-core/src/test/java/com/netflix/spinnaker/orca/pipeline/expressions/ExpressionTransformTest.java (hunk 6 above)
Comment (stuart-pollock) [reply orca-core/src/test/java/com/netflix/spinnaker/orca/pipeline/expressions/ExpressionTransformTest.java]: We think that the tests would benefit from being clear and explicit, since they serve essentially as documentation for the code.

File: orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/expressions/ExpressionTransform.java [hunk 9]
```diff
@@ -270,21 +270,31 @@ static String escapeSimpleExpression(String expression) {
...
+      String namespace = p.getNamespace();
+      for (ExpressionFunctionProvider.FunctionDefinition f : p.getFunctions()) {
         if (!f.getParameters().isEmpty() && f.getParameters().get(0).getType() == Execution.class) {
-          expressionAwareFunctions.add(f.getName());
+          String functionName = f.getName();
+          if (namespace != null) {
+            String registeredFunction = namespace + "_" + functionName;
+            String nameSpaceFunction = namespace + "." + functionName;
+            expression = expression.replaceAll("#" + nameSpaceFunction + "\\(", "#" + registeredFunction + "(");
```
[the remark this answers is not in the document]
Comment (stuart-pollock) [reply orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/expressions/ExpressionTransform.java]: We feel that the consistency of using ‘.’ instead of ‘_’ as a delimiter in the incoming SpEL expression is consistent with the use of ‘.’ as a delimiter in Javascript (for example), and will thus lead to a better user experience.  Having the pattern of using ‘.’ in Deck (and in pipeline JSON), then transforming the ‘.’ into ‘_’ under the covers in Orca feels like a small technical and conceptual price to pay for a syntax which will pay dividends in the future.

[the remark this answers is not in the document]
Comment (robzienert) [reply orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/expressions/ExpressionTransform.java]: Seems reasonable to me. Can we get rid of `_`? So there is a true One Way to do things?

[the remark this answers is not in the document]
Comment (stuart-pollock) [reply orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/expressions/ExpressionTransform.java]: Unfortunately, I do not know of a way to tell the SpEL parser that the namespacing dots should not be considered a delimiter which introduces another token.

[the remark this answers is not in the document]
Comment (ajordens) [reply orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/expressions/ExpressionTransform.java]: My preference is to only support one way of using a namespace'd fn and it sounds like that needs to be `_` delimitated (rather than have to worry about converting back and forth between implementation and documentation).

Will merge after.

[the remark this answers is not in the document]
Comment (stuart-pollock) [reply orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/expressions/ExpressionTransform.java]: There is already some precedent for using . to delimit some scope at https://github.com/spinnaker/deck/blob/0333b67defc62ae77cf5e48096ce7ea4715c1a5b/app/scripts/modules/core/src/widgets/spelText/SpelAutocompleteService.ts#L54..L55.  I think that changing those names without understanding its impact is beyond the scope of our namespacing change.  Changing our implementation to support _ would make this inconsistent.

[the remark this answers is not in the document]
Comment (stuart-pollock) [reply orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/expressions/ExpressionTransform.java]: We have decided to remove namespacing from our change, in lieu of @marchello2000 's initial suggestion of `cfServiceKey` instead of either `cloudfoundry_serviceKey` or `cloudfoundry.serviceKey`.  If `_` is going to be used for namespacing, then nothing would really prevent us from using `_` as part of the function name in Orca.  This invites the question of why we would use namespaces, since there are now two ways to define a function which is exposed in Deck with `_` contained within its name.

File: orca-core/src/test/java/com/netflix/spinnaker/orca/pipeline/expressions/functions/ServiceKeyExpressionFunctionProviderTest.java (hunk 8 above)
Comment (claymccoy) [reply orca-core/src/test/java/com/netflix/spinnaker/orca/pipeline/expressions/functions/ServiceKeyExpressionFunctionProviderTest.java]: Is it okay to mix Groovy and Kotlin in the test section of a subproject? One reason that we are avoiding adding Groovy, even in tests, is to clear the way to use more Kotlin in the future and there are apparently issues with mixing Groovy and Kotlin.
Also, there are a lot of ways to write Kotlin tests and none of them are as succinct as Spock. Can you point to ideal examples? We typically use Kotlin, JUnit, assert4j, and mockito, but I've seen variance in the Spinnaker projects.

Comment (ajordens) [reply orca-core/src/test/java/com/netflix/spinnaker/orca/pipeline/expressions/functions/ServiceKeyExpressionFunctionProviderTest.java]: I believe it is fine ... I also suspect we should be more strict on what ends up living in a `-core` artifact.
